### PR TITLE
Updated Surface Tension Equations and Examples

### PR DESCRIPTION
**Pull request by vikaskurapatibat**
Added Surface Tension equations, fixed some, added new. Added new examples and added new schemes to the existing ones.

### CODE REVIEW

File: pysph/examples/surface_tension/capillary_wave.py [hunk 1]
```diff
@@ -0,0 +1,264 @@
...
+        integrator = PECIntegrator(liquid=TransportVelocityStep(),
+                                   gas=TransportVelocityStep())
+        solver = Solver(
+            kernel=kernel, dim=dim, integrator=integrator,
+            dt=dt, tf=tf, adaptive_timestep=False)
+        return solver
+
+    def add_user_options(self, group):
+        choices = ['morris', 'tvf', 'adami_stress', 'adami', 'shadloo']
+        group.add_argument(
```
Comment (prabhuramachandran) [inline pysph/examples/surface_tension/capillary_wave.py]: This seems useless as you are not using schemes anywhere?

File: pysph/examples/surface_tension/capillary_wave.py [hunk 2]
```diff
@@ -0,0 +1,264 @@
...
+                        dest='gas', sources=['liquid', 'gas']),
+                    SurfaceForceAdami(
+                        dest='liquid', sources=['liquid', 'wall', 'gas']),
+                    SurfaceForceAdami(
+                        dest='gas', sources=['liquid', 'wall', 'gas']),
+                    SolidWallNoSlipBC(dest='liquid', sources=['wall'], nu=nu1),
+                    SolidWallNoSlipBC(dest='gas', sources=['wall'], nu=nu2),
+                ]),
+                ]
+        return adami_stress_equations
```
Comment (prabhuramachandran) [inline pysph/examples/surface_tension/capillary_wave.py]: So this only works for the Adami stress equations?

File: pysph/sph/surface_tension.py [hunk 3]
```diff
@@ -19,6 +19,198 @@
...
+
+    def loop(self, d_au, d_av, d_idx, d_m, DWIJ, d_pi00, d_pi01, d_pi10,
+             d_pi11, s_pi00, s_pi01, s_pi10, s_pi11, d_V, s_V, s_idx):
+        s2 = s_V[s_idx]*s_V[s_idx]
+        f1 = (d_pi00[d_idx]/(d_V[d_idx]*d_V[d_idx])+s_pi00[s_idx]/s2)
+        f2 = (d_pi01[d_idx]/(d_V[d_idx]*d_V[d_idx])+s_pi01[s_idx]/s2)
+        f3 = (d_pi10[d_idx]/(d_V[d_idx]*d_V[d_idx])+s_pi10[s_idx]/s2)
+        f4 = (d_pi11[d_idx]/(d_V[d_idx]*d_V[d_idx])+s_pi11[s_idx]/s2)
+        d_au[d_idx] += (DWIJ[0]*f1 + DWIJ[1]*f2)/d_m[d_idx]
+        d_av[d_idx] += (DWIJ[0]*f3 + DWIJ[1]*f4)/d_m[d_idx]
```
Comment (prabhuramachandran) [inline pysph/sph/surface_tension.py]: So this will work only in 2D, can you fix this to work in 3D also?

File: pysph/sph/surface_tension.py [hunk 4]
```diff
@@ -19,6 +19,198 @@
...
+            d_pi01[d_idx] = -factor*d_cx[d_idx]*d_cy[d_idx]
+            d_pi10[d_idx] = -factor*d_cx[d_idx]*d_cy[d_idx]
+            d_pi11[d_idx] = (-d_cy[d_idx]*d_cy[d_idx] +
+                             (mod_gradc2)/self.d)*factor
+            d_N[d_idx] = 1.0
+        else:
+            d_pi00[d_idx] = 0.0
+            d_pi01[d_idx] = 0.0
+            d_pi10[d_idx] = 0.0
+            d_pi11[d_idx] = 0.0
```
Comment (prabhuramachandran) [inline pysph/sph/surface_tension.py]: Same here, only 2d?

File: pysph/sph/surface_tension.py [hunk 5]
```diff
@@ -444,7 +636,90 @@ def loop(self, d_idx, d_au, d_av, d_aw, d_kappa,
         d_aw[d_idx] += tmp * d_cz[d_idx]
 
 
+class AdamiReproducingDivergence(Equation):
+    r"""Reproducing divergence approximation Eq. (20) in [A10] to
```
Comment (prabhuramachandran) [inline pysph/sph/surface_tension.py]: Just make sure that [A10] is defined.

File: pysph/examples/surface_tension/capillary_wave.py (hunk 2 above)
Comment (vikaskurapatibat) [reply pysph/examples/surface_tension/capillary_wave.py]: Sir,

Yes, this problem has varying density and various viscosity. Other schemes implemented won't be able to solve this problem.

File: pysph/examples/surface_tension/capillary_wave.py (hunk 1 above)
Comment (vikaskurapatibat) [reply pysph/examples/surface_tension/capillary_wave.py]: Removed that part of the code and sending another pull request. I'll send another pull request with the create_equations part sometime.

File: pysph/sph/surface_tension.py (hunk 3 above)
Comment (vikaskurapatibat) [reply pysph/sph/surface_tension.py]: Sir,

I will need to formulate the whole matrix and rewrite the formulas. I'll do it when I get some free time for sure.

File: pysph/sph/surface_tension.py (hunk 4 above)
Comment (vikaskurapatibat) [reply pysph/sph/surface_tension.py]: Same as above.

File: pysph/sph/surface_tension.py (hunk 5 above)
Comment (vikaskurapatibat) [reply pysph/sph/surface_tension.py]: It's defined in the beginning of the codes.

File: pysph/sph/surface_tension.py (hunk 3 above)
Comment (prabhuramachandran) [reply pysph/sph/surface_tension.py]: Perhaps it is easier to just setup a matrix and do a matrix multiply, there are some simple routines that will do this for you in `pysph.sph.wc.linalg`.

Comment (vikaskurapatibat) [reply pysph/sph/surface_tension.py]: This has been fixed to work in 3D also now. Please take a look

File: pysph/sph/surface_tension.py (hunk 4 above)
Comment (vikaskurapatibat) [reply pysph/sph/surface_tension.py]: This has been fixed to work for 3D also. Please take a look

File: pysph/sph/surface_tension.py [hunk 6]
```diff
@@ -500,72 +820,225 @@ def loop(self, d_idx, s_idx, d_V, s_V, d_rho, s_rho,
...
-        d_wij_sum[d_idx] += RIJ * xijdotdwij * Vj
-
-    def post_loop(self, d_idx, d_kappa, d_wij_sum):
-        # normalize the curvature estimate
-        d_kappa[d_idx] /= d_wij_sum[d_idx]
-        d_kappa[d_idx] *= self.dim
+            d_ddelta[d_idx] = 1./one_mod_gradc
+
+
+def return_equations(fluids, solids, scheme, rho0, p0, c0, b, factor1, factor2,
```
Comment (prabhuramachandran) [inline pysph/sph/surface_tension.py]: I don't like this name as it is not sufficiently descriptive.  Perhaps call it `get_surface_tension_equations`?  Also this function should have sufficient documentation in the docstring in order to explain what it does.

File: pysph/sph/surface_tension.py [hunk 7]
```diff
@@ -500,72 +820,225 @@ def loop(self, d_idx, s_idx, d_V, s_V, d_rho, s_rho,
...
+            equations.append(MomentumEquationViscosity(dest=i, sources=fluids,
+                                                       nu=nu))
+            equations.append(ShadlooYildizSurfaceTensionForce(dest=i,
+                                                              sources=None,
+                                                              sigma=sigma))
+            if len(solids) != 0:
+                equations.append(SolidWallNoSlipBC(dest=i, sources=solids,
+                                                   nu=nu))
+        Equations.append(Group(equations))
+    else:
```
Comment (prabhuramachandran) [inline pysph/sph/surface_tension.py]: What scheme is the else case for?

File: pysph/sph/surface_tension.py [hunk 8]
```diff
@@ -500,72 +820,225 @@ def loop(self, d_idx, s_idx, d_V, s_V, d_rho, s_rho,
...
-        # normalize the curvature estimate
-        d_kappa[d_idx] /= d_wij_sum[d_idx]
-        d_kappa[d_idx] *= self.dim
+            d_ddelta[d_idx] = 1./one_mod_gradc
+
+
+def return_equations(fluids, solids, scheme, rho0, p0, c0, b, factor1, factor2,
+                     nu, sigma, d, epsilon, gamma, real=False):
+    if scheme == 'tvf':
+        Equations = []
```
Comment (prabhuramachandran) [inline pysph/sph/surface_tension.py]: Do not use Equations and equations as variables, it is too confusing.  Also use uppercase/CamelCase only for class names not for variables.

Comment (prabhuramachandran) [reply pysph/sph/surface_tension.py]: Perhaps use `result` for the `Equations`.

File: pysph/sph/surface_tension.py [hunk 9]
```diff
@@ -13,11 +13,240 @@
...
+        d_au[d_idx] = 0.0
+        d_av[d_idx] = 0.0
+        d_aw[d_idx] = 0.0
+
+    def loop(self, d_idx, d_V, d_au, d_av, d_aw, s_V, d_p, s_p, DWIJ, s_idx,
+             d_m, R2IJ, XIJ, EPS, VIJ, d_nu, s_nu):
+        factor = 2.0*d_nu[d_idx]*s_nu[s_idx]/(d_nu[d_idx] + s_nu[s_idx])
+        V_i = 1/(d_V[d_idx]*d_V[d_idx])
+        V_j = 1/(s_V[s_idx]*s_V[s_idx])
+        dwijdotrij = (DWIJ[0]*XIJ[0]+DWIJ[1]*XIJ[1]+DWIJ[2]*XIJ[2])
```
Comment (prabhuramachandran) [inline pysph/sph/surface_tension.py]: Please leave space between `+` and `-` operators for readability.

File: pysph/examples/surface_tension/capillary_wave.py [hunk 10]
```diff
@@ -0,0 +1,257 @@
...
+    def create_solver(self):
+        kernel = QuinticSpline(dim=2)
+        integrator = PECIntegrator(liquid=TransportVelocityStep(),
+                                   gas=TransportVelocityStep())
+        solver = Solver(
+            kernel=kernel, dim=dim, integrator=integrator,
+            dt=dt, tf=tf, adaptive_timestep=False)
+        return solver
+
+    def create_equations(self):
```
Comment (prabhuramachandran) [inline pysph/examples/surface_tension/capillary_wave.py]: Can this not also use your `return_equations`?

File: pysph/examples/surface_tension/equilibrium_rod.py [hunk 11]
```diff
@@ -0,0 +1,170 @@
+import numpy as np
+import os
+
+from pysph.sph.surface_tension import return_equations
```
Comment (prabhuramachandran) [inline pysph/examples/surface_tension/equilibrium_rod.py]: As noted below, please change the name of this.

File: pysph/sph/surface_tension.py (hunk 6 above)
Comment (vikaskurapatibat) [reply pysph/sph/surface_tension.py]: Done. Please check now.

File: pysph/sph/surface_tension.py (hunk 7 above)
Comment (vikaskurapatibat) [reply pysph/sph/surface_tension.py]: The else is for Morris Scheme. The function returns the Morris scheme by default if nothing is chosen.

File: pysph/sph/surface_tension.py (hunk 8 above)
Comment (vikaskurapatibat) [reply pysph/sph/surface_tension.py]: Done. Please check now.

File: pysph/examples/surface_tension/capillary_wave.py (hunk 10 above)
Comment (vikaskurapatibat) [reply pysph/examples/surface_tension/capillary_wave.py]: Sir, here the physical properties like c0, rho0 are different for different fluid arrays. I didn't put that functionality in the return_equations. Also, as this is working for only one scheme, I didn't make it use the return_equations.

File: pysph/sph/surface_tension.py (hunk 9 above)
Comment (vikaskurapatibat) [reply pysph/sph/surface_tension.py]: Done. Please check now.

File: pysph/examples/surface_tension/equilibrium_rod.py (hunk 11 above)
Comment (vikaskurapatibat) [reply pysph/examples/surface_tension/equilibrium_rod.py]: Done. Please check.

File: pysph/sph/surface_tension.py [hunk 12]
```diff
@@ -500,72 +821,230 @@ def loop(self, d_idx, s_idx, d_V, s_V, d_rho, s_rho,
...
-
-        # accumulate the contributions
-        d_kappa[d_idx] += nijdotdwij * Vj
-        d_wij_sum[d_idx] += RIJ * xijdotdwij * Vj
-
-    def post_loop(self, d_idx, d_kappa, d_wij_sum):
-        # normalize the curvature estimate
-        d_kappa[d_idx] /= d_wij_sum[d_idx]
-        d_kappa[d_idx] *= self.dim
+    This function returns the required equations for the multiphase
```
Comment (prabhuramachandran) [inline pysph/sph/surface_tension.py]: Can you please elaborate the parameters, supported schemes.  Also describe the parameters properly, for example like this: https://github.com/pypr/pysph/blob/master/pysph/sph/scheme.py#L885

**Comment by prabhuramachandran** [comment]
Thanks for all the changes!  Will merge this now.

**Comment by vikaskurapatibat** [comment]
Thank you for the support, sir.

**Pull request merged**